ct: Support transactions

Lazin opened the pull request:
This PR adds transactions support to cloud topics.
There are several changes here:
* New end to end ducktape test that tries to use cloud topics with transactions;
* Necessary fix for the `cloud_topic_partition` (the metadata wasn't propagated from L0 metadata batches into the `raft_data` batches in fetch request);
* The actual change that enables transactions - the `rm_stm` treats placeholder batches similarly to `raft_data`.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->


* none

Lazin committed:
rptest: Add e2e cloud topics with transactions test

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68153-->
<!--0b323dd2-e505-490e-b3a4-737142d70099-->
<!--c3c17a1f-5b46-4127-9ae6-cbb9e68b5d46-->
<!--title-->
## Retry command for [Build#68153](https://buildkite.com/redpanda/redpanda/builds/68153)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5a02fd0f-a214-4112-9e84-5a7c05a3dd66-->
<!--TEST RESULTS-->
<!--e4abf21b-96c5-4550-b522-8ed92de49bef-->
<!--TEST RESULTS-->
<!--6719f323-a44b-4104-843b-b6bdb2a120f5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68153">build#68153</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68153#0197bbdd-b79d-4a9d-b254-e631222bcb90 | FAIL          | 0/21     | The test has failed across all retries |
| EndToEndCloudTopicsTxTest | test_write    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68153#0197bbe3-f4fc-482d-8e1d-6d47c1a9191e | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68179">build#68179</a></summary>


| test_class                   | test_method             | test_arguments                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-----------------------------|:------------------------|:-------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage      | {"cleanup_policy": "compact,delete"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68179#0197c178-a068-4db0-a25f-35089e3723c9 | FLAKY         | 19/21    | upstream reliability is '98.9010989010989'. current run reliability is '90.47619047619048'. drift is 8.42491 and the allowed drift is set to 50. The test should PASS |
| CloudTopicsTest              | test_reconciler_uploads | {"cloud_storage_type": 2}            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68179#0197c178-a067-488c-a884-6fe72016ec84 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
| CloudTopicsTest              | test_reconciler_uploads | {"cloud_storage_type": 2}            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68179#0197c17a-fa80-44ea-b095-c6f6319c061c | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
| CloudTopicsTest              | test_reconciler_uploads | {"cloud_storage_type": 1}            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68179#0197c178-a068-45b8-81c1-c62878f0efbb | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
| CloudTopicsTest              | test_reconciler_uploads | {"cloud_storage_type": 1}            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68179#0197c17a-fa80-455f-bd7f-acba5c7d6c50 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68204">build#68204</a></summary>


| test_class                         | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-----------------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68204#0197c24f-6d5c-4594-8013-2b09fe7c0542 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

Lazin committed:
ct: Propagate L0 metadata in fetch request

The cloud_topic_partition now propagates the metadata from the L0
metadat batch (dl_placeholder) to the raft_data batch returned to the
client in fetch request.

The placeholder batch contains the correct offset and term and some
other information required by the protocol. The data uploaded to S3 in
L0 objects contains record batch headers but these headers are not
initialized properly because the offsets are not allocated at the moment
when the L0 object is uploaded. The offsets are allocated when the
placeholder batches are replicated. The cloud_topic_partition now
combines this information with the user data in L0 objects to respond to
the fetch request.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Support transactions in cloud-toipcs

This commit makes rm_stm treat placeholder batches as raft_data batches
so they could participate in transactions.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

rockwotj reviewed (dismissed):
Looks good to me.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68179-->
<!--e36c73e2-b385-4112-8b4a-bee5df96b5c3-->
<!--08d7c6a5-d2d9-4cfa-8e22-680cda9b6ef9-->
<!--39c7b487-0356-44af-8d6f-05afde4de0fb-->
<!--c4b721f6-0dca-4e3d-9473-96c39dcb4bee-->
<!--title-->
## Retry command for [Build#68179](https://buildkite.com/redpanda/redpanda/builds/68179)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
```
<!--/body-->

rockwotj commented:
@Lazin @dotnwat should we remove this test for now?

https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/tests/cloud_topics_test.py

Lazin committed:
ct: Add dl_stm_command to the list of translated batches

The batch type is used by the reconciler which is enabled. There is a
test that checks it so it makes sense to keep reconciler to avoid code
rot even though the L1 metastore will change.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin commented:
> @Lazin @dotnwat should we remove this test for now?

Forgot about that. I think we should keep the reconciler enabled to avoid the code rot. I just added the batch type that it uses to the list of translated batches. In the near future when we will have metastore we will no longer replicate this batch using the user partition.

rockwotj commented:
Sounds fine, I'm not sure if we're going to want to keep the old reconciler around long term. I am assuming this reconciliation process will follow a domain NTP around and we'll likely want it to have a different lifecycle there.

Lazin commented:
> Sounds fine, I'm not sure if we're going to want to keep the old reconciler around long term. I am assuming this reconciliation process will follow a domain NTP around and we'll likely want it to have a different lifecycle there.

I think that a lot of the reconciler is reusable in either case. It's not very complicated though. It's not a problem to rewrite it.

bharathv commented on src/v/kafka/data/cloud_topic_partition.cc:
```diff
@@ -74,6 +108,12 @@ static model::record_batch make_placeholder_batch(
 
     builder.set_producer_identity(hdr.producer_id, hdr.producer_epoch);
+    if (hdr.attrs.is_control()) {
+        builder.set_control_type();
```
whats the data flow for control batches in this case? what creates them?

rockwotj replied:
https://github.com/redpanda-data/redpanda/blob/c2752256867cf2abc4b301b1d30e29bc98550dad/src/v/kafka/server/group.cc#L1924-L1930

rockwotj replied:
Well maybe that is only for consumer offsets topic

bharathv replied:
ya pretty sure thats unrelated.. control batches are created by the system (not user),.. AFAICT this code path of placeholder batch creation is directly from what user provides to produce (via  cloud_topic_partition::replicate(()), so I'm curious what sets these control bits.

rockwotj replied:
I saw no references to the control bits being set in franz-go, so I don't think they are set by clients either.

We really need to improve our validation in the produce path so we don't have to keep asking these questions and wondering about it.

bharathv replied:
I'm 100% certain they are not set by clients. In a regular transactions produce path they coexist with the data batches and are created by rm stm to track transaction boundaries. Since the same stm handles both replication of data and control batches, it does some sanity checks before creating batches in the log (eg: whether the data produced by client belongs to any active transactions etc).

I guess my question more was along the lines of how this PR ties L0 placeholder batches with control batches created by rm_stm (my initial impression after looking at the PR is probably more code is needed to get them both working together but tbh I don't have a good insight into this l0 code, so I could be missing something).

rockwotj replied:
Oh yes there are certainly data batches getting their headers overwritten here. It would be better to only do this merging/header overwrite if the batch is a placeholder batch

Lazin replied:
Here the batch we're creating just replaces the raft data batch (with different type and payload). I don't think it'll have the control bit set. I just tried to copy all fields here. We should probably just copy the header and modify some fields.

rockwotj replied:
you're assuming everything you get back is a placeholder batch. It is not here. There are going to be control batches from the rm stm that we should pass through and not mutate the header of. It's probably a noop, but I agree with Bharath it's more of a coincidence everything works out

Lazin replied:
I think we will not get any batches from rm stm here. The dataflow is like this

```
* kafka produce
* cloud_topic_partition
* batcher
* cloud_topic_partition (generate placeholders)
* rm stm (control batches can be created here)
```

this code sees only raft_data batches

rockwotj replied:
ah right this is the produce path, not consume path
